Proposed demo dataset / output notebook / output CSV / output text

mccalluc opened the pull request:
- A demonstration of the notebook that will be a final output of the UI flow. Fix #46
- A proposal for a demo dataset. Towards #7
- A proposal for CSV and Text output formats.
- An explanation of `utf8-lossy`. Towards #42

When this is approved, it will be broken into pieces that can be generated by the user, so this is a really good time to make sure we're all happy with the target output, while it's all in one piece.

@ekraffmiller:
- Does it feel like the generated notebook would be useful for the users we're imagining?
- Are the Text and CSV formats appropriate?
- Does anyone else need to be looped in here?

@Shoeboxam 
- Is the usage of OpenDP appropriate?
- Should we add more links to the docs site, or any more explanation?
- The right accuracy conversion function to use depends on distribution in the summary, but it feels like the connection is very manual: "Look above to confirm that these match." Should I do something else?
- I believe there is some connection between delta and the cutoff level in the graph, but can you remind me what it is?

Shoeboxam reviewed (commented):
Thanks for writing this up Chuck!

mccalluc commented on demo.ipynb:
What I can do now to simplify things is move this CSV generation code into a helper script. The bar plot was useful to make sure the distribution was plausible, but doesn't need to be kept.

mccalluc commented on demo.ipynb:
Good catch!

```suggestion
    "# %pip install opendp[polars]==0.11.1\n",
```

mccalluc commented on demo.ipynb:
```suggestion
    "# Data:\n",
    "csv_path = '/tmp/demo.csv'\n",
    "max_possible_rows = 1000000 \n",
    "# Privacy unit:\n",
    "contributions = 10\n",
    "# Privacy loss:\n",
    "epsilon = 2\n",
    "weights = [4, 4, 1, 1]\n",
    "delta = 1e-7\n",
```

mccalluc commented on demo.ipynb:
> I assume this is living outside because we haven't cut a nightly yet?

For this project, I'd prefer not to use nightlies. This is supposed to model what end users should do, and I think end users should be pinning versions for stability. (If there's a better way to do this with the next release, great!)

Shoeboxam replied:
I remember this comment now. You'd literally need to specify dummy bins like "A", "B", "C", and so on, which isn't as compelling as numeric ranges.

mccalluc commented on demo.ipynb:
Just filed 
- #58

I'll probably need to ask more questions about how exactly to do that.

mccalluc commented on demo.ipynb:
Just filed
- #59

I think for now we just want to drop it from the notebook, and not bother with a long explanation for the end user?

mccalluc commented on demo.ipynb:
Thanks!

mccalluc commented on demo.ipynb:
A CSV export format was requested during planning. If you think it's not useful, then best next step might be to put it on the agenda for the next meeting?

Shoeboxam replied:
Maybe needs clarification- flattening out a nested data structure into a csv is not very user-friendly. Why is this wanted? I could totally see having one csv per query. Maybe the json could be part of a zip that also holds csvs?

mccalluc replied:
Asked on Slack, and wasn't able to get an immediate answer, but Ceilyn did file a [tracking issue](https://github.com/IQSS/dataverse-pm/issues/336#issuecomment-2417563426). I'll check with Ellen tomorrow, but my sense is that no one here has a precise picture of what this should be. Maybe the end-users will, if we ask them, but it is not considered a blocker.

To move this forward, I'm going to make a new branch from here with just the outputs, so it's clear that they are separate from the rest of the notebook.

Shoeboxam commented on demo.ipynb:
Hmm, while logger output could be seen as another medium for specifying queries, I don't think that format is very human-editable, so I'd want to avoid logger tabs. If the lower editability is not consistent with your vision for this json output, that's fine.

Shoeboxam replied:
Yeah, it would be good to spruce the package up and add to docs.opendp.org. It always seems to come up when people want to integrate with OpenDP.

Shoeboxam commented:
To answer your last question: In the grouping algorithm used here, delta corresponds to the probability of releasing a group key that is unique to an individual. Releasing any group key specific to an individual runs the risk of causing a privacy violation: say, for example, if the group keys were social security numbers. 

Bin labels aren't as apparently sensitive, but since DP always guards against worst case, the math protects them both the same. An adversary could still construct bin edges for which they know a bin singles out an individual.

ekraffmiller reviewed (commented):
This looks good to me in general, and I think the idea of a notebook is helpful for users.  I think it would be also good to get feedback from someone who represents a typical user.

mccalluc commented on demo.ipynb:
done!

mccalluc commented on demo.ipynb:
Filed follow-up issue: https://github.com/opendp/opendp/issues/2102

mccalluc commented on demo.ipynb:
Sorry, I'm not understanding exactly what needs to change. There's already delta in the context:
```
privacy_loss=dp.loss_of(epsilon=epsilon, delta=delta),
```
but if I just change `keys` to `lengths` here:
```
        ("grade_bin",): dp.polars.Margin(
            max_partition_length=max_possible_rows,
-            public_info="keys",
+            public_info="lengths",
        ),
        ("class_year_bin",): dp.polars.Margin(
            max_partition_length=max_possible_rows,
-            public_info="keys",
+            public_info="lengths",
        ),
```
It errors at
```
grade_histogram_summary = grade_histogram_query.summarize()
```
```
ValueError: unable to infer bounds
```

mccalluc commented on demo.ipynb:
alpha is a required positional parameter on `discrete_laplacian_scale_to_accuracy`: Is that not correct?

mccalluc replied:
never mind: accuracy is on the summary.

mccalluc commented:
Thanks for feedback. I've addressed some points, and narrowed the scope. Two new PRs are filed for 
- input: #61
- and output: #69

My understanding from Ceilyn is that they'll try to recruit some real users to give feedback, but we shouldn't wait for that: For now, we just need to use our own judgement.

Mike, I've addressed several of your comments, but sometimes, for instance the accuracy on means, we would need to do something more complicated now to get results, pending new features in opendp. If we were just creating a report for users, work arounds might make sense so we could give them the fullest results, but since this is intended to be a model for users, I would incline towards keeping it simple, with perhaps at most a link to an issue if we want to explain that it is coming.

Please re-review!

mccalluc commented:
Some TODOs:
- The backticks around "Context" make it a little harder to see that it's a link, so take them off.
- Pull out some of the parts of Context (like the privacy loss) and explain them in their own cells.
- On the charts, make sure the y-axis doesn't go negative. (If there a negative values, we'll still see the whiskers.

mccalluc commented:
Notebook generation has been implemented. This PR was useful in setting direction, but is no longer needed.
